File: src/main/java/com/larryhsiao/clotho/comparator/StringComparator.java
package com.larryhsiao.clotho.comparator;

import java.text.Collator;
import java.util.Comparator;

/**
 * String comparator for sorting String in non-ASCII format.
 */
public class StringComparator implements Comparator<String> {
    @Override
    public int compare(String o1, String o2) {
        return Collator.getInstance().compare(o1, o2);
    }
}

File: src/main/java/com/larryhsiao/clotho/comparator/StringComparator.kt
package com.larryhsiao.clotho.comparator

import java.text.Collator

/**
 * String comparator with reversed flag.
 *
 * This implementation should work with chinese,korean, japanese.
 */
class StringComparator(private val reversed: Boolean = false) : Comparator<String> {
    override fun compare(o1: String?, o2: String?): Int {
        return if (reversed) {
            Collator.getInstance().compare(o1, o2)
        } else {
            Collator.getInstance().compare(o2, o1)
        }
    }
}